### PR TITLE
Pivot nans fix

### LINKED ISSUE #18030: Pandas pivot_table MultiIndex and dropna=False generates all combinations of modalities instead of keeping existing one only
#### Minimal Verifiable Working Example

```python
import pandas as pd
# JSON Dump for MWVE:
txt = """[{"channelid":5069,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Cu","timestamp":1514764800000,"userfloatvalue":null},{"channelid":5069,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Cu","timestamp":1514851200000,"userfloatvalue":null},{"channelid":5069,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Cu","timestamp":1514937600000,"userfloatvalue":null},{"channelid":5069,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Cu","timestamp":1515024000000,"userfloatvalue":null},{"channelid":5119,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Cu","timestamp":1514764800000,"userfloatvalue":null},{"channelid":5119,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Cu","timestamp":1514851200000,"userfloatvalue":null},{"channelid":5119,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Cu","timestamp":1514937600000,"userfloatvalue":null},{"channelid":5119,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Cu","timestamp":1515024000000,"userfloatvalue":null},{"channelid":5120,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Pb","timestamp":1514764800000,"userfloatvalue":null},{"channelid":5120,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Pb","timestamp":1514851200000,"userfloatvalue":null},{"channelid":5120,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Pb","timestamp":1514937600000,"userfloatvalue":null},{"channelid":5120,"networkkey":"HMT","sitekey":"01MEU1","measurandkey":"Pb","timestamp":1515024000000,"userfloatvalue":null},{"channelid":5233,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Pb","timestamp":1514764800000,"userfloatvalue":null},{"channelid":5233,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Pb","timestamp":1514851200000,"userfloatvalue":null},{"channelid":5233,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Pb","timestamp":1514937600000,"userfloatvalue":null},{"channelid":5233,"networkkey":"HMT","sitekey":"01AND3","measurandkey":"Pb","timestamp":1515024000000,"userfloatvalue":null}]"""
# Load Data:
df = pd.read_json(txt)
# Filling NaN with string works as expected but downcast column types:
cross2 = df.pivot_table(index="timestamp", columns=["channelid", "networkkey", "sitekey", "measurandkey"], values="userfloatvalue", aggfunc="first", fill_value="nodata")
# Trying to pivot data using MultiIndex and keeping columns of NaN produces all combinations of modalities:
cross3 = df.pivot_table(index="timestamp", columns=["channelid", "networkkey", "sitekey", "measurandkey"], values="userfloatvalue", aggfunc="first", dropna=False)
```

Input is:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>channelid</th>
      <th>measurandkey</th>
      <th>networkkey</th>
      <th>sitekey</th>
      <th>timestamp</th>
      <th>userfloatvalue</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5069</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-01</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>5069</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-02</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5069</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-03</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>5069</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-04</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5119</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-01</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5119</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-02</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>5119</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-03</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>5119</td>
      <td>Cu</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-04</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>5120</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-01</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9</th>
      <td>5120</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-02</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>5120</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-03</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>5120</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01MEU1</td>
      <td>2018-01-04</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>5233</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-01</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>5233</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-02</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>5233</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-03</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>15</th>
      <td>5233</td>
      <td>Pb</td>
      <td>HMT</td>
      <td>01AND3</td>
      <td>2018-01-04</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

Output is (`cross3`):

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>channelid</th>
      <th colspan="4" halign="left">5069</th>
      <th colspan="4" halign="left">5119</th>
      <th colspan="4" halign="left">5120</th>
      <th colspan="4" halign="left">5233</th>
    </tr>
    <tr>
      <th>networkkey</th>
      <th colspan="4" halign="left">HMT</th>
      <th colspan="4" halign="left">HMT</th>
      <th colspan="4" halign="left">HMT</th>
      <th colspan="4" halign="left">HMT</th>
    </tr>
    <tr>
      <th>sitekey</th>
      <th colspan="2" halign="left">01AND3</th>
      <th colspan="2" halign="left">01MEU1</th>
      <th colspan="2" halign="left">01AND3</th>
      <th colspan="2" halign="left">01MEU1</th>
      <th colspan="2" halign="left">01AND3</th>
      <th colspan="2" halign="left">01MEU1</th>
      <th colspan="2" halign="left">01AND3</th>
      <th colspan="2" halign="left">01MEU1</th>
    </tr>
    <tr>
      <th>measurandkey</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Cu</th>
      <th>Pb</th>
    </tr>
    <tr>
      <th>timestamp</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2018-01-01</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2018-01-02</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2018-01-03</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2018-01-04</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

Expected output is similar to `cross2` but with `NaN` value instead of string:

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>channelid</th>
      <th>5069</th>
      <th>5119</th>
      <th>5120</th>
      <th>5233</th>
    </tr>
    <tr>
      <th>networkkey</th>
      <th>HMT</th>
      <th>HMT</th>
      <th>HMT</th>
      <th>HMT</th>
    </tr>
    <tr>
      <th>sitekey</th>
      <th>01MEU1</th>
      <th>01AND3</th>
      <th>01MEU1</th>
      <th>01AND3</th>
    </tr>
    <tr>
      <th>measurandkey</th>
      <th>Cu</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Pb</th>
    </tr>
    <tr>
      <th>timestamp</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2018-01-01</th>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
    </tr>
    <tr>
      <th>2018-01-02</th>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
    </tr>
    <tr>
      <th>2018-01-03</th>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
    </tr>
    <tr>
      <th>2018-01-04</th>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
      <td>nodata</td>
    </tr>
  </tbody>
</table>

#### Problem description

I have the need:

 - to use MultiIndex in columns even if it is overdetermined (I mean, with less levels, index is still unique) and;
 - to keep columns full of NaN because it means the channel lacks all its data.

I first had a `Memory Error` with queries about 1000 rows and 25 channels, then I reduced the amount of rows and columns, and I get the following MVWE above.

#### Expected Output

To my understanding, the following setup:

```python
cross3 = df.pivot_table(index="timestamp", columns=["channelid", "networkkey", "sitekey", "measurandkey"], values="userfloatvalue", aggfunc="first", dropna=False)
```

Should return the same as:
```python
cross2 = df.pivot_table(index="timestamp", columns=["channelid", "networkkey", "sitekey", "measurandkey"], values="userfloatvalue", aggfunc="first", fill_value="nodata")
cross2bis = cross2.replace('nodata', float('nan'))
```

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>channelid</th>
      <th>5069</th>
      <th>5119</th>
      <th>5120</th>
      <th>5233</th>
    </tr>
    <tr>
      <th>networkkey</th>
      <th>HMT</th>
      <th>HMT</th>
      <th>HMT</th>
      <th>HMT</th>
    </tr>
    <tr>
      <th>sitekey</th>
      <th>01MEU1</th>
      <th>01AND3</th>
      <th>01MEU1</th>
      <th>01AND3</th>
    </tr>
    <tr>
      <th>measurandkey</th>
      <th>Cu</th>
      <th>Cu</th>
      <th>Pb</th>
      <th>Pb</th>
    </tr>
    <tr>
      <th>timestamp</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2018-01-01</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2018-01-02</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2018-01-03</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2018-01-04</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

Without raising a `MemoryError` for reasonable amount of data.

What seems to be the problem, is the creation of all combination of level modalities which drastically increase the amount of Memory without necessity. Maybe it is a bug, maybe it is the designed behaviour. Just wanted to notice it because it has surprised me.

#### Output of ``pd.show_versions()``

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-75-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 36.4.0
Cython: None
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 5.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.1
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: 1.1.9
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.8
s3fs: None
pandas_gbq: None
pandas_datareader: None
```

### PR DESCRIPTION
**Pull request by endremborza**
this gets the ball rolling on #18030 as it is very easy to fix, but requires some decisions, mostly for structure and naming

the basic options as I see them:

0 - this PR as is (maybe some name change to this 26 character monster)

1 - Outright changing the `dropna=False` default behavior I thought would be a bad idea as it breaks backwards compatibility, but this solution is quite clunky.

2 - Change `dropna=True` behavior to satisfy the test added here, but again that is not backwards compatible, but seems reasonable.

Also, I've noticed that taking out the line
https://github.com/pandas-dev/pandas/blob/3f8c0c4f002b6bcc69e0e534cb0c502fbab00373/pandas/core/reshape/pivot.py#L177
doesn't break any of the tests, so there is that. (this helps 2)

- [x] closes #18030
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_pivot.py [hunk 1]
```diff
@@ -185,6 +185,45 @@ def test_pivot_table_dropna(self):
...
+                ("Bill", "C"),
+                ("Bill", "D"),
+                ("Tom", "A"),
+                ("Tom", "B"),
+                ("Tom", "C"),
+                ("Tom", "D"),
+            ],
+            names=["measurer", "product"],
+        )
+        tm.assert_index_equal(pv_col.columns, m)
```
Comment (WillAyd) [inline pandas/tests/reshape/test_pivot.py]: Can you construct the expected output and use `tm.assert_frame_equal(result, expected` instead? You should see that in a lot of other tests

File: pandas/tests/reshape/test_pivot.py [hunk 2]
```diff
@@ -185,6 +185,45 @@ def test_pivot_table_dropna(self):
         tm.assert_index_equal(pv_col.columns, m)
         tm.assert_index_equal(pv_ind.index, m)
 
+    def test_pivot_table_keep_nancols(self):
+        df = pd.DataFrame(
```
Comment (WillAyd) [inline pandas/tests/reshape/test_pivot.py]: Can you add the issue number as a comment here?

**Review by WillAyd — CHANGES_REQUESTED** [review]
I might not fully understand the referenced issue yet (couldn't copy / paste the example to match what was written) but I don't think adding a new keyword is appropriate here. Option 2 seems more appropriate - the existing behavior is a bug right?

**Review by jreback — CHANGES_REQUESTED** [review]
this is a non-obvious api and would some discussion as this is just api bloat AFIACT.

**Comment by jreback** [comment]
why can you not simply used the value in the ``observed`` keyword?

File: pandas/tests/reshape/test_pivot.py (hunk 2 above)
Comment (endremborza) [reply pandas/tests/reshape/test_pivot.py]: done

File: pandas/tests/reshape/test_pivot.py (hunk 1 above)
Comment (endremborza) [reply pandas/tests/reshape/test_pivot.py]: did one way

**Comment by endremborza** [comment]
Made some changes to move towards solution 2, but this requires a change in default value to observed, and it breaks 2 tests referencing #21133 

those two tests basically require what this new one would like to eliminate (keeping records where values are all nan) which was possible with a new kwarg, but without it the combinations from observed/dropna cant work it out without modifying the old test

**Comment by endremborza** [comment]
I'm sorry, is there any way forward with this?

Either past tests need to be modified and backwards compatibility broken, or the API needs to be bloated, or the whole thing abandoned.

Should I modify the tests that drop the nancols now?

**Comment by WillAyd** [comment]
@endremborza can you help summarize what behavior you are expecting from the various combinations of `dropna` / `observed` and how they differ from today? We typically don't just change the default values (note CI going red) so would have to think through options and best approach

**Comment by endremborza** [comment]
simply put, I think ```observed=True``` with ```dropna=False``` should not drop combinations that are present, but filled with nans.

simplest example I can produce:
```python
df = pd.DataFrame(
    {"number": ["one", "two", "one", "three"], 
     "letter": ["a", "a" ,"b", "b"], 
     "y": [1, 2, np.nan, np.nan],
     "x": [1, 2, 3, 4]}
)

result = df.pivot_table(values="y",
                        columns=["number", "letter"],
                        index="x",
                        observed=True, dropna=False)
```

here result is

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>number</th>
      <th colspan="2" halign="left">one</th>
      <th colspan="2" halign="left">three</th>
      <th colspan="2" halign="left">two</th>
    </tr>
    <tr>
      <th>letter</th>
      <th>a</th>
      <th>b</th>
      <th>a</th>
      <th>b</th>
      <th>a</th>
      <th>b</th>
    </tr>
    <tr>
      <th>x</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>1.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>2.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

I suppose it should be

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>number</th>
      <th colspan="2" halign="left">one</th>
      <th>three</th>
      <th>two</th>
    </tr>
    <tr>
      <th>letter</th>
      <th>a</th>
      <th>b</th>
      <th>b</th>
      <th>a</th>
    </tr>
    <tr>
      <th>x</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>1.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

However, this breaks current tests as I (and the red CI) pointed out

> also, I just noticed that if I leave out the x column and the ```index="x"``` I get a weird attribute error, that I think should not happen, but that is another issue

**Comment by WillAyd** [comment]
Thanks for the PR but I think this is stale and has some disagreement over implementation. If interested in pursuing further let's discuss more in issue before coming back to code

**Comment by endremborza** [comment]
@WillAyd I'm sorry but this has been an absolute disappointment. I like pandas and I believe in the idea of actively trying to fix your own issues with open source software, but the indecision and disinterest here has been painful.

I think I've been quite clear with how all this could be handled, but it seems nobody took the time and the effort to evaluate these options.

Calling it stale just claims that there was nothing wrong with the process, or the way this was handled, which is quite off-putting.

**Comment by jreback** [comment]
@endremborza you never responded to any of my queries above

you need to engage in discussion otherwise we cannot move forwards

**Pull request closed without merging**